File: frontend/public-transportation/src/utils/trip-link.ts
import type { GeocodeSuggestion } from '../types'

export interface SharedTrip {
  origin: GeocodeSuggestion
  destination: GeocodeSuggestion
  /** null = "leave now" — resolved when the recipient's search runs */
  departureTime: Date | null
  arriveBy: boolean
}

function parseLatLon(value: string | null): { lat: number; lon: number } | null {
  if (!value) return null
  const parts = value.split(',')
  if (parts.length !== 2) return null
  const lat = Number(parts[0])
  const lon = Number(parts[1])
  if (!Number.isFinite(lat) || !Number.isFinite(lon)) return null
  if (lat < -90 || lat > 90 || lon < -180 || lon > 180) return null
  return { lat, lon }
}

/**
 * Encode the whole journey into a URL: exact coordinates plus display names,
 * so the recipient's app restores the trip without geocoding, and the time
 * mode ("now" is omitted so it resolves at open time, not at share time).
 */
export function buildTripLink(trip: SharedTrip): string {
  const params = new URLSearchParams()
  params.set('from', `${trip.origin.lat.toFixed(5)},${trip.origin.lon.toFixed(5)}`)
  params.set('fromName', trip.origin.name)
  params.set('to', `${trip.destination.lat.toFixed(5)},${trip.destination.lon.toFixed(5)}`)
  params.set('toName', trip.destination.name)
  if (trip.departureTime) params.set('time', trip.departureTime.toISOString())
  if (trip.arriveBy) params.set('arriveBy', '1')
  return `${window.location.origin}${window.location.pathname}?${params.toString()}`
}

export function parseTripLink(params: URLSearchParams): SharedTrip | null {
  const from = parseLatLon(params.get('from'))
  const to = parseLatLon(params.get('to'))
  if (!from || !to) return null

  const originName = params.get('fromName')?.trim() || `${from.lat.toFixed(4)}, ${from.lon.toFixed(4)}`
  const destinationName = params.get('toName')?.trim() || `${to.lat.toFixed(4)}, ${to.lon.toFixed(4)}`

  // An unparseable time means the sender's "depart at" can't be honored;
  // "leave now" is the only meaningful reading of the link at that point.
  let departureTime: Date | null = null
  const timeRaw = params.get('time')
  if (timeRaw) {
    const parsed = new Date(timeRaw)
    if (!Number.isNaN(parsed.getTime())) departureTime = parsed
  }

  return {
    origin: { name: originName, lat: from.lat, lon: from.lon },
    destination: { name: destinationName, lat: to.lat, lon: to.lon },
    departureTime,
    arriveBy: params.get('arriveBy') === '1',
  }
}

File: mobile/app/src/main/java/com/automatelinux/pt/util/TripLink.kt
package com.automatelinux.pt.util

import android.net.Uri
import com.automatelinux.pt.BuildConfig
import com.automatelinux.pt.data.model.GeocodeSuggestion
import kotlinx.datetime.Instant
import java.util.Locale

/**
 * Trip links use the public web app's URL format (from/fromName/to/toName/time/arriveBy),
 * so a shared link opens the full journey in the web app for anyone, and in this app
 * when it handles the link. A missing time means "leave now", resolved when opened.
 */
object TripLink {

    data class SharedTrip(
        val origin: GeocodeSuggestion,
        val destination: GeocodeSuggestion,
        val departureTime: Instant?,
        val arriveBy: Boolean
    )

    private val baseUrl =
        if (BuildConfig.FEEDBACK_ENABLED) "https://pt.dev.ya-niv.com" else "https://pt.prod.ya-niv.com"

    fun build(
        origin: GeocodeSuggestion,
        destination: GeocodeSuggestion,
        departureTime: Instant?,
        arriveBy: Boolean,
        /**
         * Sharer's anonymous install id, carried so an install that starts from
         * this link can be credited back. Attribution only — [parse] ignores it,
         * so it can never affect which journey the recipient sees.
         */
        referrerInstallId: String? = null
    ): String {
        val builder = Uri.parse(baseUrl).buildUpon()
            .appendQueryParameter("from", formatCoords(origin.lat, origin.lon))
            .appendQueryParameter("fromName", origin.name)
            .appendQueryParameter("to", formatCoords(destination.lat, destination.lon))
            .appendQueryParameter("toName", destination.name)
        if (departureTime != null) builder.appendQueryParameter("time", departureTime.toString())
        if (arriveBy) builder.appendQueryParameter("arriveBy", "1")
        if (!referrerInstallId.isNullOrBlank()) builder.appendQueryParameter("ref", referrerInstallId)
        return builder.build().toString()
    }

    fun parse(uri: Uri?): SharedTrip? {
        if (uri == null) return null
        val from = parseCoords(uri.getQueryParameter("from")) ?: return null
        val to = parseCoords(uri.getQueryParameter("to")) ?: return null

        val originName = uri.getQueryParameter("fromName")?.trim().takeUnless { it.isNullOrEmpty() }
            ?: formatCoords(from.first, from.second)
        val destinationName = uri.getQueryParameter("toName")?.trim().takeUnless { it.isNullOrEmpty() }
            ?: formatCoords(to.first, to.second)

        // An unparseable time can only be read as "leave now" — same rule as the web app.
        val departureTime = uri.getQueryParameter("time")?.let {
            try { Instant.parse(it) } catch (_: Exception) { null }
        }

        return SharedTrip(
            origin = GeocodeSuggestion(name = originName, lat = from.first, lon = from.second),
            destination = GeocodeSuggestion(name = destinationName, lat = to.first, lon = to.second),
            departureTime = departureTime,
            arriveBy = uri.getQueryParameter("arriveBy") == "1"
        )
    }

    private fun formatCoords(lat: Double, lon: Double): String =
        String.format(Locale.US, "%.5f,%.5f", lat, lon)

    private fun parseCoords(value: String?): Pair<Double, Double>? {
        if (value == null) return null
        val parts = value.split(",")
        if (parts.size != 2) return null
        val lat = parts[0].toDoubleOrNull() ?: return null
        val lon = parts[1].toDoubleOrNull() ?: return null
        if (lat < -90 || lat > 90 || lon < -180 || lon > 180) return null
        return Pair(lat, lon)
    }
}
